docs/spec.md

```diff
@@ -173,10 +191,48 @@ values of type `tensor<T>`).
   and an **imaginary part** of the same **element type**. Supported complex
   types are `complex<f32>` (both parts are of type `f32`) and `complex<f64>`
   (both parts are of type `f64`).
-* In the future, we are also planning to introduce **quantized types** that
-  represent integer values obtained via uniform quantization of floating-point
-  values using given scales and zero points
-  ([#588](https://github.com/openxla/stablehlo/issues/588)).
+* **Quantized types** represent integer values of a **storage type** in the
+  range from `storage_min` to `storage_max` (inclusive) that correspond to
+  floating-point values of an **expressed type**. For a given integer value `i`,
+  the corresponding floating-point value `f` can be computed as
+  `f = (i - zero_point) * scale`, where `scale` and `zero_point` are called
+  **quantization parameters**.
+
+Quantization can be **per-tensor**, meaning, having one `scale` and/or
+`zero_point` for the entire tensor or can be **per-axis**, meaning, having
+multiple `scales` and/or `zero_points`, one pair per slice of a particular
+dimension `quantized_dimension`. More formally, in a tensor `t` of per-axis
+quantized element type, there are `dim(t, quantized_dimension)` slices of the
+`quantized_dimension`: `t[:, ..., 0, ..., :], t[:, ..., 1, ..., :]`, etc. All
+elements in the `i`th slice use `scales[i]` and `zero_points[i]` as their
+quantization parameters.
+
+Quantized types satisfy the following constraints:
+
+* (C1) `num_bits(storage_type) < num_bits(expressed_type)`.
+* (C2) `type(storage_min) = storage_type`.
+* (C3) `type(storage_max) = storage_type`.
+* (C4) `storage_min` and `storage_max` have the following values:
+  * If `is_signed(storage_type)` and `d = num_bits(storage_type)`:
+    * `storage_min = -2^(d-1)`, `storage_max = 2^(d-1)-1` OR
+    * `storage_min = -2^(d-1)+1`, `storage_max = 2^(d-1)-1`.
+  * If `is_unsigned(storage_type)`:
+    * `storage_min = 0`, `storage_max = 2^d - 1`.
+* (C5) `type(scale) = expressed_type`.
+* (C6) `scale > 0`.
+* (C7) `type(zero_point) = i64`.
```

Comment:
In addition to the type constraint, we need another value range constraint on `zero_point`s:

```
(C?) `storage_min <= zero_point <= storage_max`
```

We need this constraint because we must ensure that floating point zero `0.0` can always be quantized into a quantized value so that we can pad the activation properly (e.g. convolution padding).

Reply:
Yes.  "The floating point `0.0` will always lie in the range" will be implied if we add this constraint.

Reply:
Even though I can't find a good source of documentation, many math don't work if the floating point `0.0` can't be exactly quantized.  For example, in TFLite optimizer, they need some adjustment to ensure `0.0` is included:

https://github.com/tensorflow/tensorflow/blob/f93185001ff1d7b75e9966b80a144601d7d6e696/tensorflow/lite/tools/optimize/quantization_utils.cc#L74

Also, mentioned in [Quantization and Training of Neural Networks for Efficient Integer-Arithmetic-Only Inference (Jacob et al, 2017)](https://arxiv.org/pdf/1712.05877.pdf) p.3.

Reply:
Thanks @loganchien!  As per your feedback proposed edits as mentioned in https://github.com/openxla/stablehlo/pull/1352#discussion_r1157559676